src/hupper/reloader.py

```diff
@@ -113,157 +123,195 @@ def run(self):
+
+                else:  # pragma: no cover
+                    raise RuntimeError('received unknown control signal', cmd)
+
+                # done handling the packet, continue to the next one
+                # do not fall through here because it will block
+                continue
 
-        while worker.is_alive():
-            if monitor.is_changed():
-                result = WorkerResult.FILE_CHANGED
+            if not worker.is_alive():
                 break
 
-            try:
-                cmd = worker.pipe.recv(timeout=0.5)
-            except queue.Empty:
-                continue
+            signal = os.read(self.control_r, 1)
 
-            if cmd is None:
-                if worker.is_alive():
-                    # the worker socket has died but the process is still
-                    # alive (somehow) so wait a brief period to see if it
-                    # dies on its own - if it does die then we want to
-                    # treat it as a crash and wait for changes before
-                    # reloading, if it doesn't die then we want to force
-                    # reload the app immediately because it probably
-                    # didn't die due to some file changes
-                    time.sleep(reload_interval)
-
-                if worker.is_alive():
-                    logger.info('Broken pipe to server, triggering a reload.')
-                    result = WorkerResult.BROKEN_PIPE
-
-                else:
-                    logger.debug('Broken pipe to server, looks like a crash.')
+            if not signal:
+                logger.error('Control pipe died unexpectedly.')
                 break
 
-            if cmd[0] == 'reload':
-                logger.debug('Server triggered a reload.')
-                result = WorkerResult.RELOAD_REQUEST
+            elif signal == ControlSignal.SIGINT:
+                logger.info('Received SIGINT, waiting for server to exit ...')
+                result = WorkerResult.EXIT
                 break
 
-            if cmd[0] == 'watch':
-                for path in cmd[1]:
-                    monitor.add_path(path)
+            elif signal == ControlSignal.SIGHUP:
+                logger.info('Received SIGHUP, triggering a reload.')
+                result = WorkerResult.RELOAD
+                break
 
-            else:  # pragma: no cover
-                raise RuntimeError('received unknown command')
+            elif signal == ControlSignal.SIGTERM:
+                logger.info('Received SIGTERM, triggering a shutdown.')
+                result = WorkerResult.EXIT
+                break
 
-        if worker.is_alive() and shutdown_interval is not None:
-            logger.info('Gracefully killing the server.')
-            worker.kill(soft=True)
-            worker.wait(shutdown_interval)
+            elif signal == ControlSignal.SIGCHLD:
+                if not worker.is_alive():
+                    break
```

Comment:
currently on windows SIGCHLD will not occur, but it's not a big deal because we have an open pipe with the child. When the child dies, the pipe is closed and we emit a WORKER_COMMAND with a `None` packet and we're able to exit pretty quickly via the `if cmd is None` check above.